Mathlib/Algebra/Hom/Group.lean

```diff
@@ -0,0 +1,1533 @@
+infixr:25 " →* " => MonoidHom
+
+/-- `MonoidHomClass F M N` states that `F` is a type of `Monoid`-preserving homomorphisms.
+You should also extend this typeclass when you extend `MonoidHom`. -/
+-- Porting note: see porting note of `ZeroHomClass`
+@[to_additive]
+class MonoidHomClass (F : Type _) (M N : outParam (Type _)) [outParam (MulOneClass M)]
+   [outParam (MulOneClass N)] extends MulHomClass F M N, OneHomClass F M N
+#align monoid_hom_class MonoidHomClass
+
+@[to_additive]
+instance MonoidHom.monoidHomClass : MonoidHomClass (M →* N) M N where
+  coe f := f.toFun
+  coe_injective' f g h := by
+    cases f
+    cases g
+    congr
+    apply OneHom.oneHomClass.coe_injective'
+    exact h
+  map_mul := MonoidHom.map_mul'
+  map_one f := f.toOneHom.map_one'
+#align monoid_hom.monoid_hom_class MonoidHom.monoidHomClass
+
+-- Porting note: help to_additive translate. This should ideally be handled by the tactic
+attribute [to_additive AddMonoidHomClass.toZeroHomClass] MonoidHomClass.toOneHomClass
+
+-- Porting note: restore `to_additive`
+/-- Any type that is a `MonoidHomClass` can be cast into a `MonoidHom`. -/
+-- @[to_additive "Any type that is a `AddMonoidHomClass` can be cast into a `AddMonoidHom`."]
+instance [MonoidHomClass F M N] : CoeTC F (M →* N) :=
+  ⟨fun f => { toFun := f, map_one' := map_one f, map_mul' := map_mul f }⟩
+/-- Any type that is a `AddMonoidHomClass` can be cast into a `AddMonoidHom`. -/
+instance [AddZeroClass M] [AddZeroClass N] [AddMonoidHomClass F M N] : CoeTC F (M →+ N) :=
+  ⟨fun f => { toFun := f, map_zero' := map_zero f, map_add' := map_add f }⟩
+attribute [to_additive instCoeTCAddMonoidHom] instCoeTCMonoidHom
+
+-- Porting note: restore `to_additive`
+@[simp]
+theorem MonoidHom.coe_coe [MonoidHomClass F M N] (f : F) : ((f : M →* N) : M → N) = f := rfl
+#align monoid_hom.coe_coe MonoidHom.coe_coe
+@[simp]
+theorem AddMonoidHom.coe_coe [AddZeroClass M] [AddZeroClass N] [AddMonoidHomClass F M N] (f : F) :
+  ((f : M →+ N) : M → N) = f := rfl
+#align add_monoid_hom.coe_coe AddMonoidHom.coe_coe
+attribute [to_additive AddMonoidHom.coe_coe] MonoidHom.coe_coe
+
+@[to_additive]
+theorem map_mul_eq_one [MonoidHomClass F M N] (f : F) {a b : M} (h : a * b = 1) : f a * f b = 1 :=
+  by rw [← map_mul, h, map_one]
+#align map_mul_eq_one map_mul_eq_one
+
+@[to_additive]
+theorem map_div' [DivInvMonoid G] [DivInvMonoid H] [MonoidHomClass F G H]
+  (f : F) (hf : ∀ a, f a⁻¹ = (f a)⁻¹) (a b : G) : f (a / b) = f a / f b :=
+  by rw [div_eq_mul_inv, div_eq_mul_inv, map_mul, hf]
+#align map_div' map_div'
+
+/-- Group homomorphisms preserve inverse. -/
+@[simp, to_additive "Additive group homomorphisms preserve negation."]
+theorem map_inv [Group G] [DivisionMonoid H] [MonoidHomClass F G H]
+  (f : F) (a : G) : f a⁻¹ = (f a)⁻¹ :=
+  eq_inv_of_mul_eq_one_left <| map_mul_eq_one f <| inv_mul_self _
+#align map_inv map_inv
+
+-- Porting note: redundant `simp` lemma
+#noalign map_mul_inv
```

Comment:
I don't understand how this could be redundant? It probably needs to have `simp` removed, but the lemma needs to be there.